### PR TITLE
Pointer instructions

### PR DESCRIPTION
**Pull request by gianbelinche**
Closes https://github.com/lambdaclass/era_vm/issues/7
This PR adds `ptr.add`, `ptr.sub`, `ptr.pack` and `ptr.shrink`opcodes

### CODE REVIEW

File: src/op_handlers/ptr_add.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    state::VMState,
+    value::FatPointer,
+    Opcode,
+};
+
+pub fn _ptr_add(vm: &mut VMState, opcode: &Opcode) {
+    let (pointer, diff, src0) = ptr_operands_read(vm, opcode, "ptr_add");
+    let (new_offset, overflow) = pointer.offset.overflowing_add(diff);
+    if overflow {
+        panic!("Offset overflow in ptr_add");
```
Comment (fkrause98) [inline src/op_handlers/ptr_add.rs:12]: Let's add a todo here for when we properly implement the vm's panic routine.

File: src/value.rs [hunk 2]
```diff
@@ -2,8 +2,80 @@ use u256::U256;
...
+    fn default() -> Self {
+        Self {
+            value: U256::zero(),
+            is_pointer: false,
+        }
+    }
+}
+
+impl TaggedValue {
+    pub fn new(value: U256, is_pointer: bool) -> Self {
```
Comment (fkrause98) [inline src/value.rs:28]: I think this function is redundant since we already have the other 2 below.

File: programs/add_ptr_stack.zasm [hunk 3]
```diff
@@ -0,0 +1,16 @@
+	.text
+	.file	"add.zasm"
+	.globl	__entry
+__entry:
+.func_begin0:
+	; sets r1=FatPointer from test
+	add 5,r0,r2
+	add r0,r0,stack+=[1]
+	ptr.add r1,r0,stack=[0]
```
Comment (fkrause98) [inline programs/add_ptr_stack.zasm:9]: How does ptr.add work? I dont' remember seeing this on the spec.

File: src/value.rs (hunk 2 above)
Comment (gianbelinche) [reply src/value.rs]: Done

File: src/op_handlers/ptr_add.rs (hunk 1 above)
Comment (gianbelinche) [reply src/op_handlers/ptr_add.rs:12]: Done

File: programs/add_ptr_stack.zasm (hunk 3 above)
Comment (gianbelinche) [reply programs/add_ptr_stack.zasm:9]: ptr.add receives two inputs, the first one is a Fat pointer, with the following fields:
`{offset, page, start, len }`
The second input is a number usually called diff, and the result of ptr.add is a new pointer `{offset + diff, page, start, len}`
You can find the specification on page 205 on the EraVM ISA Specification

File: src/op_handlers/ptr_shrink.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    ptr_operator::{ptr_operands_read, ptr_operands_store},
+    state::VMState,
+    value::FatPointer,
+    Opcode,
+};
+
+pub fn _ptr_shrink(vm: &mut VMState, opcode: &Opcode) {
+    let (pointer, diff, src0) = ptr_operands_read(vm, opcode, "ptr_shrink");
+    let (new_len, overflow) = pointer.len.overflowing_sub(diff);
+    if overflow {
```
[the remark this answers is not in the document]
Comment (gianbelinche) [reply src/op_handlers/ptr_shrink.rs:11]: On page 213 it says on the PtrShrink specification
> Panics
> 3. Shrinking underflows.

It contradict itself then

File: src/op_handlers/ptr_pack.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+};
+
+pub fn _ptr_pack(vm: &mut VMState, opcode: &Opcode) {
+    let (src0, src1) = address_operands_read(vm, opcode);
+
+    if !src0.is_pointer || src1.is_pointer {
+        panic!("Invalid operands for ptr_pack");
+    }
+
+    if src1.value & U256::from_str_radix("0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF", 16).unwrap()
```
[the remark this answers is not in the document]
Comment (gianbelinche) [reply src/op_handlers/ptr_pack.rs]: Done

File: src/ptr_operator.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+) -> (FatPointer, u32, TaggedValue) {
+    let (src0, src1) = address_operands_read(vm, opcode);
+
+    if !src0.is_pointer || src1.is_pointer {
+        panic!("Invalid operands for {}", opcode_name);
+    }
+
+    let pointer = FatPointer::decode(src0.value);
+    if src1.value > MAX_OFFSET_FOR_ADD_SUB {
+        panic!("Src1 too large for {}", opcode_name);
```
[the remark this answers is not in the document]
Comment (gianbelinche) [reply src/ptr_operator.rs]: Done

File: src/op_handlers/ptr_shrink.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (juan518munoz) [reply src/op_handlers/ptr_shrink.rs:11]: Correct! I would keep the panic then.

**Review by juan518munoz — APPROVED** [review]
LGTM

**Review by IAvecilla — APPROVED** [review]
LGTM!

**Pull request merged**